daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/SAXInfosetInputter.scala

```diff
@@ -121,38 +124,45 @@ class SAXInfosetInputter(
   }
 
   override def hasNext(): Boolean = {
-    if (endDocumentReceived) false
-    else if (!nextEvent.isEmpty) true
-    else {
-      val event = this.resume(unparseContentHandler, Try(currentEvent))
-      copyEvent(source = event, dest = nextEvent)
+    val nextIndex = currentIndex + 1
+    if (endDocumentReceived) {
+      // if the current Element is EndDocument, then there is no valid next
+      false
+    } else if (batchedInfosetEvents != null && batchedInfosetEvents.lift(nextIndex).nonEmpty
+      && !batchedInfosetEvents(nextIndex).isEmpty) {
+      // if the next element exists and is nonEmpty
+      true
+    }  else  {
+      // there is no nextEvent or it was empty, but we still have no EndDocument. So load the next
+      // batch from the contentHandler
+      returnedInfosetEvent(0) = batchedInfosetEvents(currentIndex)
+      batchedInfosetEvents = this.resume(unparseContentHandler, returnedInfosetEvent)
+      // we reset the index once we receive a new batch, but we don't use the first element from
+      // this point on
+      currentIndex = 0
```

Comment:
I think this comment could be a little confusing. It maybe implies that we're doing something wasteful? The reason this must be set to 0 is because hasNext isn't allowed to change what we are looking at. Calling resume took the thing we were last looking at and made it the first thing in the array. So the currentIndex must now be changed to look at the zeroth thing. Only when next() is called are we allowed to actually increment currentIndex so we are actually looking at the next element.

Reply:
I'll update the comment, but I decided not to the copy, since once we context switch back to the inputter, we immediately call next, which clears whatever is at the currentIndex (i.e 0) away.

Reply:
You're correct in practice, but in theory daffodil should be allowed to call hasNext without immediately calling next. For example, it could call hasNext(), getLocalName(), and then next(). In this case, getLocalName() still must refer the the same event bevor hasNext() was called. We could someday change the daffodil code so that happens. In which case, I think this won't work properly. I think we do need to do a copy in the ContentHandler (see on of my other comments) to allow this potential behavior.

Reply:
That makes sense. Updated!